alfresco-search/src/test/java/org/alfresco/repo/search/impl/lucene/analysis/SynonymAwareStopFilterTest.java

```diff
@@ -0,0 +1,235 @@
+    {
+        File stopwordsFile = testFolder.newFile("stopwords.txt");
+        File synonyms = testFolder.newFile("synonyms.txt");
+
+        FileUtils.writeLines(stopwordsFile, stopwords);
+        FileUtils.writeLines(synonyms, asList("OOW,Out of Warranty","transfer Phone Number,Port Number"));
+
+        System.setProperty("solr.solr.home", stopwordsFile.getParentFile().getAbsolutePath());
+        resourceLoader = new SolrResourceLoader();
+
+        analyzer = new Analyzer()
+        {
+            @Override
+            protected TokenStreamComponents createComponents(String fieldName)
+            {
+                try
+                {
+                    Tokenizer tokenizer = new StandardTokenizerFactory(Collections.emptyMap()).create();
+
+                    Map<String, String> synonymsConfig = new HashMap<>();
+                    synonymsConfig.put("synonyms", "synonyms.txt");
+                    synonymsConfig.put("ignoreCase", "true");
+
+                    SynonymGraphFilterFactory synonymsFactory = new SynonymGraphFilterFactory(synonymsConfig);
+                    synonymsFactory.inform(resourceLoader);
+                    TokenStream synonymsFilter = synonymsFactory.create(tokenizer);
+
+                    Map<String, String> stopFilterConfig = new HashMap<>();
+                    stopFilterConfig.put("words", "stopwords.txt");
+                    stopFilterConfig.put("ignoreCase", "true");
+
+                    SynonymAwareStopFilterFactory stopWordsFactory = new SynonymAwareStopFilterFactory(stopFilterConfig);
+                    stopWordsFactory.inform(resourceLoader);
+
+                    TokenStream sinkStream = stopWordsFactory.create(synonymsFilter);
+
+                    return new TokenStreamComponents(tokenizer, sinkStream);
+                }
+                catch (Exception exception)
+                {
+                    throw new IllegalArgumentException(exception);
+                }
+            }
+        };
+    }
+
+    /**
+     * In case we have no synonyms and no stopwords then the input should be left untouched.
+     */
+    @Test
+    public void noSynonmsNoStopwords() throws Exception
+    {
+        final String text = "no synonyms and no stopwords here";
+        final List<String> expected = asList(text.split("\\W"));
+        this.assertAnalysisCorrectness(text, expected);
+    }
+
+    /**
+     * If a stream doesn't contain any SYNONYM token, the custom stop filter behaves like its superclass (StopFilter)
+     */
+    @Test
+    public void noSynonymsOnlyStopwords() throws Exception
+    {
+        final String text = "No synonyms detection. However we have two stopwords of and my which will be removed";
+        final List<String> expectedTokens =
+                stream(text.split("\\W|\\p{Punct}"))
+                        .map(String::trim)
+                        .filter(token -> !token.isEmpty())
+                        .filter(token -> !stopwords.contains(token))
```

Comment:
Rather than filter like this it might be simpler (to read) to just start from the expected output tokens in a string.

Reply:
Ok, I replaced the stream with a simple list of expected tokens. I agree, it introduces some redundancy but it more readable.